srcts/src/events/shinyEvents.ts

```diff
@@ -2,35 +2,250 @@ import type { InputBinding } from "../bindings/input/inputBinding";
 import type { OutputBindingAdapter } from "../bindings/outputAdapter";
 import type { EventPriority } from "../inputPolicies/inputPolicy";
 import type { ErrorsMessageValue } from "../shiny/shinyapp";
+import type { EvtFn } from "./jQueryEvents";
+import $ from "jquery";
+
+// This class implements a common interface for all Shiny events, and provides a
+// layer of abstraction between the Shiny event and the underlying jQuery event
+// object. We use a new class, rather than extending JQuery.Event, because
+// jQuery.Event is an old function-style class. Each Event class has a
+// corresponding ShinyEvent interface that describes the event object that is
+// emitted. At the end of this file, we extend JQuery's `on()` method to
+// associate the ShinyEvent interfaces with their corresponding event string.
+class EventBase {
+  event: JQuery.Event;
+
+  constructor(type: string) {
+    this.event = $.Event(type);
+  }
+
+  triggerOn(
+    el: HTMLElement | JQuery<HTMLElement> | typeof document | null
+  ): void {
+    $(el || window.document).trigger(this.event);
+  }
+
+  isDefaultPrevented(): boolean {
+    return this.event.isDefaultPrevented();
+  }
+}
 
 interface ShinyEventCommon extends JQuery.Event {
   name: string;
-  value: unknown;
-  el: HTMLElement | null;
+  value: any;
```

Comment:
Why `any` here? Typically `unknown` is safer.

Reply:
tl/dr; My current understanding is to use `any` as we can not change the type of `unknown` to another value in a sub class.

It comes from https://en.wikipedia.org/wiki/SOLID's  The [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle):
> Functions that use pointers or references to base classes must be able to use objects of derived classes without knowing it.

I broke this rule all over the place in the original conversion within `rstudio/shiny`. (This rule alone warrants a rewrite of most of the `unknown` types to `any` or Generic types.)

For example: Let say `ShinyEventCommon` has `value: unknown`. Let's have `NickEventInputChanged` extend `ShinyEventCommon`. Even if you know `NickEventInputChanged`'s value has type `NickValue`, the `value` must satisfy it's parent's class `value` type of `unknown`.  This forces `NickEventInputChanged` to then have a `value: unknown`.




------------------------------

But after writing this all out, I am thinking of `any` vs `never`. The `never` type can not be recovered. The `unknown` type requires some type checking but can be eventually be used (it just requires that you type check it).

I think it'd be good to have a quick discussion at the end of standup today.

Reply:
Could we make this all go away if we used a generic type on `value` where `T extends any` and `value: T`. Then we don't need to worry what the type as it is always defined by the Generic type.